### PR TITLE
router: fix streaming shadows bug

### PR DESCRIPTION
**Pull request by paul-r-gall**
Commit Message: router: fix streaming shadows bug
Additional Description: Fixes a potential use-after-free when streaming_shadows is enabled. This PR was cleared with the envoy security committee.
Risk Level: low
Testing: unit tests

### CODE REVIEW

File: test/common/router/router_test.cc [hunk 1]
```diff
@@ -4705,6 +4705,51 @@ TEST_P(RouterShadowingTest, NoShadowForConnect) {
   router_->onDestroy();
 }
 
+TEST_P(RouterShadowingTest, LahDeeDah) {
```
Comment (adisuissa) [inline test/common/router/router_test.cc:4708]: Please use a better test name :)
and consider adding a one-line comment explaining this test.

**Review by adisuissa — COMMENTED** [review]
Thanks for fixing this!
One small comment, but otherwise LGTM.

File: test/common/router/router_test.cc (hunk 1 above)
Comment (paul-r-gall) [reply test/common/router/router_test.cc]: Whoopsie, sorry. Done.

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!
Assigning Greg for non-Google review
/assign @ggreenway

**Comment by paul-r-gall** [comment]
MSAN failure looks like a flake.

**Comment by adisuissa** [comment]
Previous MSAN failure was due to another PR.
@ggreenway can you PTAL?

**Comment by adisuissa** [comment]
@wbpcode can you please review this PR as a !Google reviewer?

**Review by ggreenway — APPROVED** [review]
LGTM

**Pull request merged**